http-server/src/main/java/org/triplea/server/access/ApiKeyGenerator.java

```diff
@@ -0,0 +1,48 @@
+package org.triplea.server.access;
+
+import com.google.common.hash.Hashing;
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import lombok.Builder;
+import org.triplea.http.client.ApiKey;
+import org.triplea.lobby.server.db.dao.ApiKeyDao;
+import org.triplea.lobby.server.db.dao.UserJdbiDao;
+
+// TODO: Project#12 This is copy/pasted from lobby sub-project
+@Builder
+public class ApiKeyGenerator implements Supplier<ApiKey> {
+
+  @Nonnull private final ApiKeyDao apiKeyDao;
+  @Nonnull private final UserJdbiDao userDao;
+  @Nonnull private final Supplier<ApiKey> keyMaker;
+
+  /**
+   * Generates a random string value that is extremely difficult to guess or crack. <br>
+   * This key generator is intended to have a pretty fast runtime. <br>
+   * <br>
+   * To generate:<br>
+   * (1) We concatenate a UUID and random number together to get a lot of randomness.<br>
+   * (2) The concatenated random value is hashed to prevent guessing the next value. This makes it
+   * difficult to crack the seed value being used.<br>
+   * (3) Finally, we attach nano epoch to make the API key even more difficult to guess. <br>
+   */
+  public static Supplier<ApiKey> createKeyMaker() {
+    return () -> ApiKey.of(randomHash() + System.nanoTime());
+  }
+
+  private static String randomHash() {
+    return Hashing.sha512()
+        .hashString(UUID.randomUUID() + String.valueOf(Math.random()), StandardCharsets.UTF_8)
+        .toString();
+  }
```

Comment:
I removed the Math.random(), please see the javadoc that this is a copy/paste: https://github.com/triplea-game/triplea/pull/5311/files#diff-031f3c07836a4fe1aedf4678ac4800c7R13

This is a roll-your-own random string, I did not find a library to generate an API key and overall this does not seem like it's that bad. We are not rolling our own crypto as the hashing is done by libraries, the only thing extra now that the random number is removed is we are suffixing with a nano time stamp.

I have trouble seeing the nano time stamp as hurting, an attacker would need to not only crack the SHA512 but also guess the nano time stamp. The time stamp could only really hurt if it somehow would yield information about the seed value used to generate the UUID.

I don't think we should use `SecureRandom` directly though, it seems like we'd expose a whole lot of details about the seed and converting to and from byte arrays.

When it comes to the nano timestamp, I think it could be argued both ways, unless it hurts, I'm thinking we're good enough here and should avoid the code churn.

Reply:
Yeah, I read the javadoc but I didn't realize where exactly it was coming from, just checked makes a lot more sense now.
Anyways, here's what I was thinking:
I believe this system is unecessarily complex and I'm concerned we come back at this in the future and ask ourselves what exactly is actually required from this code and what kind of special goals we tried to achieve with it.
I personally think something like
```java
String randomHash() {
  final var random = new SecureRandom();
  final var bytes = new byte[512 / 8]; // Same amount of bits like SHA-512
  random.nextBytes(bytes);
  return com.google.common.io.BaseEncoding.base16().encode(bytes);
}
```
makes it super clear the only thing that matters is making the string unguessable. And avoids the pseudo-randomness step by SHA-512.

In the end it doesn't really matter. because the UUID part alone should make the whole thing difficult enough to guess already. But I'd be happy to create a PR for this soon™ if you don't have any strong feelings against my approach.

Reply:
This part I'm not a fan of:   
```
final var bytes = new byte[512 / 8]; // Same amount of bits like SHA-512
```

Given we've dropped to the byte level, that very much strikes me as 'roll-your-own'.

If we're not happy with the timestamp append, then removing that and we are left with:
```
    return Hashing.sha512()hashString(UUID.randomUUID()).toString()
```

The current version appends a timestamp to the above, it's pretty straight forward.

Reply:
> I believe this system is unecessarily complex and I'm concerned we come back at this in the future and ask ourselves what exactly is actually required from this code and what kind of special goals we tried to achieve with it.

Well stated, though seeing a byte array raises this exact concern for me.

Reply:
> because the UUID part alone should make the whole thing difficult enough to guess already.

The concern i would have is if someone can determine the seed value then you could determine the UUID. Hence the nano timestamp makes it so even if you can figure out that seed value, you have to then know the nano second when the api key was generated. That makes it really strong IMO.

I wonder a bit if the nano time stamp should be appended before hashing vs after. I don't think it actually even matters because if you've cracked the UUID generation then you're still left guessing the nano value either way.

I'm not sure if there is that much difference between base16 and sha512, though sha512 for me raises no questions, it's a very strong hash.

Reply:
Hmm I think my real problem is the hashing. In my opinion it's completely unecessary, instead of brute-forcing all 2^512 combinations, it would be "easier" to brute-force 2^128 combinations of UUIDs and apply the hashing function to it.

So I am wondering: Is there anything speaking agains just using `UUID.randomUUID().toString()`?
If you're paranoid about the way those UUIDs are generated you could even use an assertion:
```java
final var id = UUID.randomUUID();
assert id.version() == 4: "A random UUID is required";
return id.toString();
```
From a functionality perspective it's the same I suggested, with the only difference being the output format and the fact that it "only" generates 128 random bits. From the ease of use perspective it's arguably simpler than what we currently have.

What do you think?

Reply:
Also while nanoTime does make the string harder to guess, it is somewhat deterministic, so I wouldn't bet that it is impossible to guess the right nano time under special circumstances

Reply:
So what I'm saying is that I'd suggest just using UUID without anything else

Reply:
Also have a look at the UUID javadocs, interesting read:
https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/UUID.html#version()

Reply:
The assertion to verify how the UUID is generated is interesting, I just wonder if we can deterministically control that. It would not be fun for the assertion to fail in different systems.

Another perspective to think about, it's not just a matter of the DB being leaked and someone trying to crack the API keys, but also someone trying to brute force the API keys themselves.

> so I wouldn't bet that it is impossible to guess the right nano time under special circumstances

I'm curious what conditions you would be thinking of. For sure someone might be able to narrow down a time range, yet still, narrowing that range down to an exact nano IMO would be really impressive.

Reply:
I'm thinking `sha512(UUID + nano)` is pretty 'cheap' in terms of developer WTFs and is damn secure. UUID by itself is already quite secure. The nano is there essentially as a salt in case somehow the UUID sequence were to become compromised.

Overall, I do think stepping back is probably the thing to do and I'm thinking this:
-> API key should be UUID + nano
-> Store the API key as `sha512(UUID+nano)`

The nano then acts as a true salt. `sha512`. If the DB were to be leaked, because sha512 is a fast hash algorithm, there is some room to brute force. Nano then comes in as a salt, even with a row creation timestamp, we multiply the possibilities by a factor of 100k.

@RoiEXLab thoughts?

Reply:
Just to comment, bcrypt is not necessarily appropriate as the validation is done more frequently compared to passwords. We can increase entropy to compensate for speed of hash algorithm.

Reply:
@RoiEXLab I think leaving as-is for now is okay. Storing key as a hash in DB is a good thing to do, also need to consolidate the two generators, and there may be some more changes in api-key (I'm thinking we perhaps will want to store IP address with the api-key as well if not already). I'll follow up with these updates later as part of Project#12